<a href="https://colab.research.google.com/github/akshayvalle1989-source/Rebounce-Applied-AI-and-Analytics-Codes/blob/main/Pyhton301_akshay.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Data Wrangling with Numpy and Pandas**

**The dataset · Online Retail:**
A UK-based online gift retailer, every transaction between 1 December 2010 and 9 December 2011. Many of its
customers are wholesalers, which is why some orders are enormous and others are a single item. There are 541,909
rows and 8 columns.

# **2. Import Libraries**

In [4]:
import pandas as pd
import numpy as np

# **3. Load Dataset**

In [5]:
url = "https://huggingface.co/datasets/aarav912/online-retail/resolve/main/online_retail.csv"
df = pd.read_csv(url)

# **4. Understand the Data**

In [6]:
df.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,2010-12-01 08:26:00,7.65,17850.0,United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 08:26:00,4.25,17850.0,United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,2010-12-01 08:28:00,1.85,17850.0,United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,2010-12-01 08:28:00,1.85,17850.0,United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,2010-12-01 08:34:00,1.69,13047.0,United Kingdom


In [7]:
df.shape

(541909, 8)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [9]:
df.columns


Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

# **5. Clean the Data**

In [10]:
df.isnull().sum()

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0


In [11]:
df.duplicated().sum()

np.int64(5268)

columns = "Description","CustomerID" need cleaning


In [12]:
df[["Description", "CustomerID"]].head()

,Description,CustomerID
0,WHITE HANGING HEART T-LIGHT HOLDER,17850.0
1,WHITE METAL LANTERN,17850.0
2,CREAM CUPID HEARTS COAT HANGER,17850.0
3,KNITTED UNION FLAG HOT WATER BOTTLE,17850.0
4,RED WOOLLY HOTTIE WHITE HEART.,17850.0


In [13]:
df = df.dropna(subset=['CustomerID'])
df['CustomerID'].shape

(406829,)

In [14]:
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])
df.shape

(406829, 8)

In [15]:
df.drop_duplicates().dropna()
df.shape

(406829, 8)

In [16]:
high=df[df["Quantity"] > 0]
print(high.shape)
high.head()

(397924, 8)


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


# **6. Create Sales Column**

In [17]:
df['Sales'] = df['Quantity'] * df['UnitPrice']
print(df.shape)
df.head()


(406829, 9)


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Sales
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34


# **7. NumPy Analysis**

In [18]:
sales = df['Sales'].values
mean =round(np.mean(sales),2)
median = round(np.median(sales),2)
minimum = round(np.min(sales),2)
maximum = round(np.max(sales),2)
std_dev = round(np.std(sales),2)
print(f"Mean: {mean}")
print(f"Median: {median}")
print(f"Minimum: {minimum}")
print(f"Maximum: {maximum}")
print(f"Standard Deviation: {std_dev}")

Mean: 20.4
Median: 11.1
Minimum: -168469.6
Maximum: 168469.6
Standard Deviation: 427.59


Mean > Median:
The data is right-skewed (positively skewed)
There are outliers on the high end pulling the mean upward,although most sales are lower,few high sales are present


# **8. Pandas Filtering**

In [19]:
count = (df['Sales'] > 500).sum()
print(f"Transactions with Sales > 500: {count}")


Transactions with Sales > 500: 913


In [20]:
top_10 = df.sort_values('Sales', ascending=False).head(10)
print(top_10)

       InvoiceNo StockCode                          Description  Quantity  \
540421    581483     23843          PAPER CRAFT , LITTLE BIRDIE     80995   
61619     541431     23166       MEDIUM CERAMIC TOP STORAGE JAR     74215   
222680    556444     22502       PICNIC BASKET WICKER 60 PIECES        60   
173382    551697      POST                              POSTAGE         1   
348325    567423     23243  SET OF TEA COFFEE SUGAR TINS PANTRY      1412   
52711     540815     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
160546    550461     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
421601    573003     23084                   RABBIT NIGHT LIGHT      2400   
52709     540815    85123A   WHITE HANGING HEART T-LIGHT HOLDER      1930   
160542    550461    85123A   WHITE HANGING HEART T-LIGHT HOLDER      1930   

               InvoiceDate  UnitPrice  CustomerID         Country      Sales  
540421 2011-12-09 09:15:00       2.08     16446.0  United Kingdom  168469

In [21]:
count = (df['Country'] == 'Germany').sum()
print(f"Transactions from Germany: {count}")

Transactions from Germany: 9495


In [22]:
count2=(df['Country']== 'Netherlands').sum()
print(f"transactions from Netherlands: {count2}")


transactions from Netherlands: 2371


In [23]:
qty = df[df['Quantity'] > 1000]
print(f"Transactions with Quantity > 1,000: {len(qty)}")
print(qty.head)

Transactions with Quantity > 1,000: 106
<bound method NDFrame.head of        InvoiceNo StockCode                        Description  Quantity  \
4850      536809     84950     ASSORTED COLOUR T-LIGHT HOLDER      1824   
4945      536830     84077  WORLD WAR 2 GLIDERS ASSTD DESIGNS      2880   
4946      536830     21915             RED  HARMONICA IN BOX       1400   
6365      536890    17084R              ASSORTED INCENSE PACK      1440   
16435     537659     22189            CREAM HEART CARD HOLDER      1008   
...          ...       ...                                ...       ...   
533812    581115     22413    METAL SIGN TAKE IT OR LEAVE IT       1404   
534952    581175     23084                 RABBIT NIGHT LIGHT      1440   
540070    581458     22197                     POPCORN HOLDER      1500   
540071    581459     22197                     POPCORN HOLDER      1200   
540421    581483     23843        PAPER CRAFT , LITTLE BIRDIE     80995   

               InvoiceDate  U

# **9. GroupBy & Aggregation**

In [24]:
country_summary = df.groupby('Country').agg(
    Number_of_Transactions=('Sales', 'count'),
    Total_Sales=('Sales', 'sum'),
    Average_Sales=('Sales', 'mean')
).reset_index()
print(country_summary)

                 Country  Number_of_Transactions  Total_Sales  Average_Sales
0              Australia                    1259   137077.270     108.877895
1                Austria                     401    10154.320      25.322494
2                Bahrain                      17      548.400      32.258824
3                Belgium                    2069    40910.960      19.773301
4                 Brazil                      32     1143.600      35.737500
5                 Canada                     151     3666.380      24.280662
6        Channel Islands                     758    20086.290      26.499063
7                 Cyprus                     622    12946.290      20.813971
8         Czech Republic                      30      707.720      23.590667
9                Denmark                     389    18768.140      48.247147
10                  EIRE                    7485   250285.220      33.438239
11    European Community                      61     1291.750      21.176230

In [25]:
country_summary = country_summary.sort_values('Total_Sales', ascending=False)
print(country_summary.head())

           Country  Number_of_Transactions  Total_Sales  Average_Sales
35  United Kingdom                  361878  6767873.394      18.702086
23     Netherlands                    2371   284661.540     120.059696
10            EIRE                    7485   250285.220      33.438239
14         Germany                    9495   221698.210      23.348943
13          France                    8491   196712.840      23.167217


In [26]:
country_summary = country_summary.sort_values('Average_Sales', ascending=False)
print(country_summary.head())

        Country  Number_of_Transactions  Total_Sales  Average_Sales
23  Netherlands                    2371    284661.54     120.059696
0     Australia                    1259    137077.27     108.877895
19        Japan                     358     35340.62      98.716816
31       Sweden                     462     36595.91      79.211926
9       Denmark                     389     18768.14      48.247147


In [27]:
country_summary = country_summary.sort_values('Total_Sales', ascending=False)
print(country_summary.head(1))

           Country  Number_of_Transactions  Total_Sales  Average_Sales
35  United Kingdom                  361878  6767873.394      18.702086


In [28]:
region_lookup = {
"United Kingdom": "Europe",
"France": "Europe",
"Germany": "Europe",
"Spain": "Europe",
"Netherlands": "Europe",
"Belgium": "Europe",
"Switzerland": "Europe",
"Portugal": "Europe",
"Australia": "Oceania",
"Japan": "Asia",
"USA": "Americas",
}
region_df = pd.DataFrame(list(region_lookup.items()), columns=['Country', 'Region'])

print(region_df)

           Country    Region
0   United Kingdom    Europe
1           France    Europe
2          Germany    Europe
3            Spain    Europe
4      Netherlands    Europe
5          Belgium    Europe
6      Switzerland    Europe
7         Portugal    Europe
8        Australia   Oceania
9            Japan      Asia
10             USA  Americas


# **10. Merge / Region Analysis**

In [29]:
merged= pd.merge(country_summary, region_df, on="Country", how="left")
merged


,Country,Number_of_Transactions,Total_Sales,Average_Sales,Region
0,United Kingdom,361878,6767873.394,18.702086,Europe
1,Netherlands,2371,284661.540,120.059696,Europe
2,EIRE,7485,250285.220,33.438239,NaN
3,Germany,9495,221698.210,23.348943,Europe
4,France,8491,196712.840,23.167217,Europe
5,Australia,1259,137077.270,108.877895,Oceania
6,Switzerland,1877,55739.400,29.696004,Europe
7,Spain,2533,54774.580,21.624390,Europe
8,Belgium,2069,40910.960,19.773301,Europe
9,Sweden,462,36595.910,79.211926,NaN


In [30]:
merged.fillna("Other")

,Country,Number_of_Transactions,Total_Sales,Average_Sales,Region
0,United Kingdom,361878,6767873.394,18.702086,Europe
1,Netherlands,2371,284661.540,120.059696,Europe
2,EIRE,7485,250285.220,33.438239,Other
3,Germany,9495,221698.210,23.348943,Europe
4,France,8491,196712.840,23.167217,Europe
5,Australia,1259,137077.270,108.877895,Oceania
6,Switzerland,1877,55739.400,29.696004,Europe
7,Spain,2533,54774.580,21.624390,Europe
8,Belgium,2069,40910.960,19.773301,Europe
9,Sweden,462,36595.910,79.211926,Other


In [31]:
report=merged.groupby("Region").agg(
    Total_Sales=("Total_Sales", "sum"),
 ).reset_index()
print(report)


     Region  Total_Sales
0  Americas     1730.920
1      Asia    35340.620
2    Europe  7651430.734
3   Oceania   137077.270


In [32]:
sorted_report=report.sort_values("Total_Sales", ascending=False)
print(sorted_report)

     Region  Total_Sales
2    Europe  7651430.734
3   Oceania   137077.270
1      Asia    35340.620
0  Americas     1730.920


# **11. Answer the Questions:**

1. Which country generated the highest total sales?
**United Kingdom**




2. Which country had the highest average transaction value? **The Netherlands**

3. Which region performed best? **Europe**

4. What is one interesting pattern you noticed in the data?
*  The United Kingdom: dominates with 361,878 transactions (by far the Largest volume), but has the lowest average sales at $18.70 per transaction.
* The Netherlands fewer transactions but higher average sales 6.4x of UK


5. What is one data-quality problem you ran into, and how did you handle it?


* I discovered that my region lookup dictionary didn't cover all countries in the dataset. Several European countries like EIRE and Sweden were classified as 'Other' instead of 'Europe,' which skewed regional analysis. I handled it by expanding the lookup dictionary, re-merging the data, and verifying there were no remaining unmapped values.




# **12. Final Findings:**

**One important finding — the single thing you would tell someone who had thirty seconds:**


1.The United Kingdom generates nearly 95% of all European volume yet has the lowest average transaction value at $18.70, suggesting you may have one massive, low-value customer base rather than high-value enterprise clients.



**One comparison — two countries, two regions, or the mean against the median:**


1.The Netherlands' average sales per transaction—6.4× higher than the UK. At the regional level, Europe dominates with 99.7% of total sales, while the Americas contributes only $1,730 despite being a developed market.



**One observation about the data itself its shape, its spread, its skew:**

1. Data is heavily right-skewed, with the majority of transactions clustered at low values but pulled upward by outliers. This explains why your mean and median differ significantly the median is a much more reliable picture of a "typical" customer transaction than the mean.



**One limitation — what you had to clean, and what that means for your numbers:**


1. had to clean inconsistent country mappings (EIRE, Scandinavia) and reclassify unmapped regions, plus you spotted suspicious data like USA's $5.95 average.

